go/inst/instance_topology_dao.go

```diff
@@ -579,14 +581,172 @@ func StartReplicationUntilMasterCoordinates(instanceKey *InstanceKey, masterCoor
+		}
+		return replicaInstance, nil
+	}
+
+	// New logic: If EnforceExactSemiSyncReplicas or RecoverLockedSemiSyncMaster are set, we enable semi-sync only if the
+	// given replica instance is in the list of replicas to have semi-sync enabled (according to the priority).
+	masterInstance, err := ReadTopologyInstance(&replicaInstance.MasterKey)
+	if err != nil {
+		return replicaInstance, log.Errore(err)
+	}
+	replicas, err := ReadReplicaInstances(&replicaInstance.MasterKey)
+	if err != nil {
+		return replicaInstance, log.Errore(err)
+	}
+
+	possibleSemiSyncReplicas, asyncReplicas, excludedReplicas := ClassifyAndPrioritizeReplicas(replicas, &replicaInstance.Key)
+	actions := DetermineSemiSyncReplicaActions(possibleSemiSyncReplicas, asyncReplicas, masterInstance.SemiSyncMasterWaitForReplicaCount, masterInstance.SemiSyncMasterClients, config.Config.EnforceExactSemiSyncReplicas)
+
+	log.Debugf("semi-sync: %+v: determining whether to enable rpl_semi_sync_slave_enabled", replicaInstance.Key)
+	log.Debugf("semi-sync: master = %+v, master semi-sync wait count = %d, master semi-sync replica count = %d", replicaInstance.MasterKey, masterInstance.SemiSyncMasterWaitForReplicaCount, masterInstance.SemiSyncMasterClients)
+	LogSemiSyncReplicaAnalysis(possibleSemiSyncReplicas, asyncReplicas, excludedReplicas, actions, func(s string, a ...interface{}) { log.Debugf(s, a...) })
+
+	// TODO This is a little odd. We sometimes get actions for replicas that are not us. If we do not take this action
+	// TODO then we'll absolutely have a MasterWithTooManySemiSyncReplicas event following this recovery. We could prevent this
+	// TODO by just executing all actions, but we're in the scope of `replicaInstance` here, so..... idk
+	for replica, enable := range actions {
+		if replica.Key.Equals(&replicaInstance.Key) {
+			log.Infof("semi-sync: %s: setting rpl_semi_sync_slave_enabled: %t", &replicaInstance.Key, enable)
+			return SetSemiSyncReplica(&replicaInstance.Key, enable)
+		}
+	}
+
+	log.Infof("semi-sync: %+v: no action taken", &replicaInstance.Key)
+	return replicaInstance, nil
+}
+
+// ClassifyAndPrioritizeReplicas takes a list of replica instances and classifies them based on their semi-sync priority, excluding
+// replicas that are downtimed or down. It furthermore prioritizes the possible semi-sync replicas based on SemiSyncPriority, PromotionRule
+// and hostname (fallback).
+func ClassifyAndPrioritizeReplicas(replicas []*Instance, includeNonReplicatingInstance *InstanceKey) (possibleSemiSyncReplicas []*Instance, asyncReplicas []*Instance, excludedReplicas []*Instance) {
+	// Filter out downtimed and down replicas
+	possibleSemiSyncReplicas = make([]*Instance, 0)
+	asyncReplicas = make([]*Instance, 0)
+	excludedReplicas = make([]*Instance, 0)
+	for _, replica := range replicas {
+		if replica.IsDowntimed || !replica.IsLastCheckValid || (!replica.Key.Equals(includeNonReplicatingInstance) && !replica.ReplicaRunning()) {
```

Comment:
This logic is confusing to read: too many negatives or negatives (and negatives or negatives)...
Let's convert this `if-else` into a `switch` statement, and break down the specific `if` clause into multiple `case`s.

Reply:
I updated this to make it a little more readable, though I left the if/else. I don't quite know how to make this a switch case that is more readable. Maybe you can help:

```golang
for _, replica := range replicas {
	isReplicating := replica.Key.Equals(includeNonReplicatingInstance) || replica.ReplicaRunning()
	if !replica.IsLastCheckValid || !isReplicating {
		excludedReplicas = append(excludedReplicas, replica)
	} else if replica.SemiSyncPriority == 0 || replica.IsDowntimed {
		asyncReplicas = append(asyncReplicas, replica)
	} else {
		possibleSemiSyncReplicas = append(possibleSemiSyncReplicas, replica)
	}
}
```